### PR TITLE
Refactoring of socket related functions to a separate socket_common.c

### PR DESCRIPTION
**Pull request by Jakio815**
This PR should be merged with lf-lang/lingua-franca#2449.

## Motivation
The original codebase had socket-related functionality directly embedded in various parts of the application. This tight coupling led to code duplication and hindered modularity. In this refactor:
- Socket-related code was separated into a dedicated file: socket_common.c and socket_common.h.
- Code duplication was removed, making the codebase cleaner and easier to maintain.
- The refactor enhances modularity, allowing for easier extension, especially for supporting additional protocols and future integration of security features.
```diff
- NOTE: The primary purpose of this PR is refactoring.
- There is almost no functionality change except for the minor logic change mentioned at the end of the document.
```

## Overview of Approach
Three major functions were introduced to handle the socket-related operations:
- `create_TCP_server()`
  - This function creates a TCP server socket, binds it to a port, and enables listening for incoming connections. It handles retries in case of failure and provides a robust way to manage the server’s socket.

- `accept_rti_socket() / accept_federate_socket()`
  - This function waits for incoming connection requests on a given socket, accepting them when a valid connection is received. It ensures that temporary errors like EAGAIN or EWOULDBLOCK are handled properly and retries the connection attempt when necessary.

- `connect_to_socket()`
  - This function attempts to establish a TCP or UDP connection to a remote server. It retries the connection if necessary, cycling through different port numbers and respecting a connection timeout.

These functions are used both in the RTI (Real-Time Infrastructure) server and the federate components, enhancing reusability and simplifying the network-related logic in these modules.

### New integrated function: `create_TCP_server()` `create_UDP_server()`
`void create_TCP_server(uint16_t port, int* final_socket, uint16_t* final_port);`
`void create_UDP_server(uint16_t port, int* final_socket, uint16_t* final_port);`

**Arguments**
- `port`: The port number to bind the server socket to.
  - If a non-zero value is specified, it attempts to bind to that port.
  - If 0 is specified, it starts binding at `DEFAULT_PORT`.
- `final_socket`: A pointer to an integer where the server socket file descriptor will be stored upon success.
- `final_port`: A pointer to a uint16_t where the successfully bound port number will be stored.
- `increment_port_on_retry`: Boolean to retry port increment.

**Return Values**
- The function modifies final_socket and final_port to reflect the created server socket and bound port.
- On failure, it retries binding up to `PORT_BIND_RETRY_LIMIT` times. If all attempts fail, the function prints an error and exits.

**Function Description**
This function creates a TCP/UDP server socket and binds it to a port, enabling it to listen for incoming connections. The process includes:
- Port Handling:
  - If a specific port is provided, the function attempts to bind to it.
  - If the specified port is 0 and `increment_port_on_retry` is true, it starts from `DEFAULT_PORT` (15045) and increments the port number up to `MAX_NUM_PORT_ADDRESSES` times, cycling back to DEFAULT_PORT if necessary.
  - If the specified port is 0 and `increment_port_on_retry` is false, it means the federate is trying to bind a port assigned by the OS. 
- Retries: Retries binding up to `PORT_BIND_RETRY_LIMIT` times with a `PORT_BIND_RETRY_INTERVAL` delay between attempts when cycling ports.

**Where It’s Used**
- `lf_create_server()` in `federate.c`: Initializes a server socket for federates.
- `start_rti_server()` in `rti_remote.c`: Sets up the RTI server socket for listening. Both `create_TCP_server()` and `create_UDP_server()` is called. 

### New integrated function: `accept_rti_socket()` `accept_federate_socket()`
`int accept_rti_socket(int socket);`
`int accept_federate_socket(int socket, int rti_socket);`

**Arguments**
- `socket`: The server socket file descriptor that is already set up and listening for incoming connections.
- `rti_socket`: The RTI socket descriptor is used to check if the RTI socket is still open/alive.

**Return Values**
- Returns the file descriptor for the newly accepted connection socket (int) on success.
- Returns -1 on failure, printing an appropriate error message and setting `errno`.

**Function Description**

This function blocks and waits for an incoming connection request on the given server socket:
- Connection Acceptance: Calls `accept()` to accept an incoming client connection.
- Error Handling:
  - If a temporary error (e.g., `EAGAIN`, `EWOULDBLOCK`, or `EINTR`) occurs, the function retries.
  - If a fatal error occurs (e.g., invalid socket, RTI closed), the function logs an error and exits.
- RTI Socket Check: Verifies the RTI socket (rti_socket) to ensure it is still open during the process.

**Where It’s Used**
- `lf_connect_to_federates()` in `rti_remote.c`: Waits for federates to connect to the RTI server.
- `respond_to_erroneous_connections()` in `rti_remote.c`: Accepts connections who are attempting to join the wrong federation.
- `lf_handle_p2p_connections_from_federates()` in `federate.c`: Accepts connections from federates that send messages directly to this federate (not through the RTI).

### New integrated function: `connect_to_socket()`
`int connect_to_socket(int sock, const char* hostname, int port)`

**Inputs**
- `sock`: The socket file descriptor created using `socket()`.
- `hostname`: The target hostname or IP address of the server to connect to.
- `port`: The port number to connect to. If 0 is specified, it is trying to connect to the RTI. The function iterates through a range of default ports starting at `DEFAULT_PORT`.

**Outputs**
- Returns 0 on success, indicating the connection was established.
- Returns -1 on failure, printing an error message and setting `errno`.

**Function Description**

This function attempts to establish a TCP connection to the specified hostname and port:
- Hostname Check: Uses `getaddrinfo()` to resolve the hostname to a valid network address.
- Connection Attempts:
  - If the port is 0, the function iterates through a range of ports starting from `DEFAULT_PORT`.
  - Retries periodically until `CONNECT_TIMEOUT` is reached.
- Error Handling: If a connection attempt fails, the function logs a warning and retries after a short delay `CONNECT_RETRY_INTERVAL`.

**Where It’s Used**
- `lf_connect_to_federate()` in `federate.c`: Federates use this function to establish connections with other federates.
- `lf_connect_to_rti()` in `federate.c`: Federates use this function to establish connections with the RTI server.

**Minor Logic Changed**
The original code for `lf_connect_to_rti()` and `lf_connect_to_federate()`, there was a large while loop retrying the connect() and also sending the  `MSG_TYPE_FED_IDS` and `MSG_TYPE_P2P_SENDING_FED_ID` messages respectively. Below is a simplified version.
```
 instant_t start_connect = lf_time_physical();
  while (!CHECK_TIMEOUT(start_connect, CONNECT_TIMEOUT) && !_lf_termination_executed) {
    connect();
    // Send MSG_TYPE_FED_IDS message.
  }
```
The refactored version has two while loops, one inside the `connect_to_socket()` function itself and the original while loop. Below is the new simplified version.

```
 connect_to_socket(); // The while loop inside checks timeout.
 instant_t start_connect = lf_time_physical();
  while (!CHECK_TIMEOUT(start_connect, CONNECT_TIMEOUT) && !_lf_termination_executed) {
    connect();
    // Send MSG_TYPE_FED_IDS message.
  }
```
## Follow Up
This PR yet did not handle the `shutdown()` and `close()` functions. This will be done in the next PR as soon as possible.
UPDATE: This is done in #506

### CODE REVIEW

File: core/federated/RTI/rti_remote.c
Comment (hokeun) [inline core/federated/RTI/rti_remote.c:1142]: Why is this commented out?

File: core/federated/RTI/rti_remote.c
Comment (hokeun) [inline core/federated/RTI/rti_remote.c:1591]: Can you explain this change?

File: core/federated/RTI/rti_remote.h
Comment (hokeun) [inline core/federated/RTI/rti_remote.h:25]: Why is this added?

File: include/core/federated/network/socket_common.h
Comment (hokeun) [inline include/core/federated/network/socket_common.h:232]: Can you add a new line at the end of this file?

File: core/federated/RTI/rti_remote.c
Comment (Jakio815) [reply core/federated/RTI/rti_remote.c:1142]: Oh, I should probably remove this.

Actually this part will be removed in the future. This part will be included in the accept() function, when in the next PR.

This part was refactored to remove the input argument `struct sockaddr_in* client_fd` in this function.

File: core/federated/RTI/rti_remote.c
Comment (Jakio815) [reply core/federated/RTI/rti_remote.c:1591]: Sure. Thank you for pointing this out.

The federate and RTI both initialize the `user_specified_port` as 0. 
For the RTI, when the `user_specified_port` is 0, it recognizes that there was no user input to assign a port and assigns the port to the `DEFAULT_PORT`, which is 15045.

For the federate, when the `user_specified_port` is 0, it also recognizes that there was no user input and makes the OS assign the port. 

The `create_TCP_server` gets this input argument `port`, which is actually the `user_specified_port` for both RTI and federate. To distinguish whether the RTI or federate is calling this function, I changed the initial value of the RTI's port as 1.

We need to distinguish whether the RTI or federate called the `create_TCP_server` function for two reasons.
1. As explained before, we should set the default port, which RTI and federate differ.
2. RTI has increments the port number starting from 15045, when the port binding fails. However, the federate does not increment the port number and try binding.

File: core/federated/RTI/rti_remote.h
Comment (Jakio815) [reply core/federated/RTI/rti_remote.h:25]: It was needed because the socket_type_t was moved to socket_common.h, however we do not need this anymore. I'll remove this.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:232]: Sure! Thank you for pointing this out. I'll also check other files.

File: core/federated/RTI/rti_remote.c
Comment (hokeun) [reply core/federated/RTI/rti_remote.c:1586]: Can you add this information in the comment? We want to better understand this decision later. You can even leave a link to this PR in the comment.

**Review by hokeun — APPROVED** [review]
Looks good to me, minus the missing comment that I suggested. - Please address my comment.
Overall, I think this change will reduce code duplication.

Please wait for @edwardalee 's feedback before merging.

File: include/core/federated/network/socket_common.h
Comment (edwardalee) [inline include/core/federated/network/socket_common.h:120]: It bothers me that this function uses port number 0 to mean what port number 1 means for the create_socket function. This is potentially confusing.  I think maybe a better design would to have a boolean argument `increment_port_on_retry` for both functions. Then, to get the effect of port number 1, you specify port number `DEFAULT_PORT` and `increment_port_on_retry == true`.

**Review by edwardalee — CHANGES_REQUESTED** [review]
This is looking good, but it still needs work.  I haven't reviewed everything, but my suggestions will lead to enough changes that it's better to take this in stages.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:27]: Thank you for your suggestions.

File: core/federated/RTI/rti_remote.c
Comment (Jakio815) [reply core/federated/RTI/rti_remote.c:1602]: Thank you for your suggestions.

File: include/core/federated/network/net_util.h
Comment (Jakio815) [reply include/core/federated/network/net_util.h]: Thank you. I moved it to the top of the file with the guards.

File: core/federated/network/socket_common.c
Comment (Jakio815) [reply core/federated/network/socket_common.c:174]: The `create_UDP_server()` function is currently only used for creating the clock synchronization server.
<pre>
  if (rti_remote->clock_sync_global_status >= clock_sync_on) {
    <b>create_UDP_server(rti_remote->final_port_TCP + 1, &rti_remote->socket_descriptor_UDP, &rti_remote->final_port_UDP);</b>
  }
  return rti_remote->socket_descriptor_TCP;
}
</pre>


I split these functions into two, because, I wanted to change that part's usage as `create_clock_sync_server()` when I integrate the network driver.

I think there is no reason for the clock synchronization be done by anything other than UDP. 
If we have no plans except using UDP, I can add this function in `clock-sync.c`.

So, is it worth to use other network stacks for the clock-synchronization?

Comment (edwardalee) [reply core/federated/network/socket_common.c:174]: I think clock sync really needs to use UDP, but there are currently no plans to use it for anything else.

File: core/federated/network/socket_common.c
Comment (Jakio815) [reply core/federated/network/socket_common.c]: I fixed the `listen()` to get the return value, and set `errno`.
Also fixed the create_server() function itself, and added the return value for error checking as you suggested.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h]: Thank you for explaining the reason. I just added comments.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:133]: I agree with that it could be confusing.
I reverted the changes, and set the RTI initial port back to zero. I added the `increment_port_on_retry.`

It really looks much better now!

File: core/federated/network/socket_common.c
Comment (Jakio815) [reply core/federated/network/socket_common.c]: Then would it be fine to leave this as is? I would not like to leave the `socket_type_t` as an argument, for further development.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:133]: I also cleaned up all comments about changing the port number in the code, and the PR description.

File: core/federated/network/socket_common.c
Comment (Jakio815) [reply core/federated/network/socket_common.c]: I removed the separated functions, and just left the `accept_socket()` as you suggested in [this commit](https://github.com/lf-lang/reactor-c/pull/505/commits/5a26cf5be43c7d3b0bfe8d3fe9816bc9057524d5) .

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h]: I removed the separated functions, and just left the `accept_socket()` as you suggested in [this commit](https://github.com/lf-lang/reactor-c/pull/505/commits/5a26cf5be43c7d3b0bfe8d3fe9816bc9057524d5) .

File: include/core/federated/network/socket_common.h
Comment (edwardalee) [inline include/core/federated/network/socket_common.h:130]:
```suggestion
 * Attempt to establish a TCP connection to the specified hostname
 * and port. This function uses `getaddrinfo` to resolve the hostname and retries the connection
```

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:6]: Looks right. I think this was in the old code. I'll also have a look at the other ones.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:130]: Thank you for your suggestion.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:107]: Thank you for your suggestion.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:100]: Thank you for your suggestion.

File: include/core/federated/network/socket_common.h
Comment (Jakio815) [reply include/core/federated/network/socket_common.h:117]: Thank you for your suggestion.

File: core/federated/RTI/rti_remote.h
Comment (hokeun) [reply core/federated/RTI/rti_remote.h:416]: I just confirmed that there is a new line. Resoving this comment.

File: core/federated/network/socket_common.c
Comment (Jakio815) [reply core/federated/network/socket_common.c]: @edwardalee I removed the separate functions, and just left `create_server()`. It is in this commit https://github.com/lf-lang/reactor-c/pull/505/commits/b2a5af8206f7a973d04eba8661d8978889fd2ed9.

File: include/core/federated/network/socket_common.h
Comment (edwardalee) [inline include/core/federated/network/socket_common.h:101]:
```suggestion
 * @param sock_type Type of the socket, TCP or UDP.
```

**Pull request merged**
